+++ venv/lib/python3.10/site-packages/torchgeometry/core/conversions.py
@@ -298,6 +298,9 @@
                       rmat_t[:, 0, 1] - rmat_t[:, 1, 0]], -1)
     t3_rep = t3.repeat(4, 1).t()
 
+    mask_d2 = mask_d2.float()
+    mask_d0_d1 = mask_d0_d1.float()
+    mask_d0_nd1 = mask_d0_nd1.float()
     mask_c0 = mask_d2 * mask_d0_d1
     mask_c1 = mask_d2 * (1 - mask_d0_d1)
     mask_c2 = (1 - mask_d2) * mask_d0_nd1
